### PR TITLE
[RHINENG-7074] Convert GET /hosts/{host_id_list}/tags/count endpoint from XJoin to DB

### PR DESCRIPTION
**Pull request by chambridge**
# Overview

This PR is being created to address [RHINENG-7074](https://issues.redhat.com/browse/RHINENG-7074).

* Add feature flag check to enable query via posgtres instead of xjoin
* Add a get_host_tags_list_by_id_list function using only sqlalchemy leveraging existing functions
* Add validation checks to keep behavior consistent with xjoin experience
* Update existing get_host_list_by_id_list_from_db function to allow you to specify columns to return instead of the entire model (minimizing the amount of data that needs to be extracted and returned from the DB when possible)
* Add some unit test cases that call the API with the feature flag enabled allowing testing of the postgres flow
* Update docker-compose better kafka integration
* Update run_gunicorn to use hot reloads for easier development

## PR Checklist

- [ ] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: tests/test_tags.py [hunk 1]
```diff
@@ -351,6 +353,30 @@ def test_get_host_tag_count_RBAC_denied(mq_create_four_specific_hosts, mocker, a
             find_non_culled_hosts_mock.assert_not_called()
 
 
+def test_get_tags_count_of_host_that_does_not_exist_via_db(api_get):
+    """
+    send a request for some host that doesn't exist via db only
+    """
+    url = build_tags_count_url(host_list_or_id="fa28ec9b-5555-4b96-9b72-96129e0c3336")
```
Comment (kruai) [inline tests/test_tags.py:360]: This is fine, but we generally use `generate_uuid()` when we want some arbitrary uuid

**Review by kruai — DISMISSED** [review]

File: dev.yml [hunk 2]
```diff
@@ -30,6 +31,22 @@ services:
...
+      command: "bash -c 'echo Waiting for Kafka to be ready... && \
+                        cub kafka-ready -b kafka:9092 1 20 && \
+                        kafka-topics --create --if-not-exists --topic platform.inventory.host-ingress --bootstrap-server kafka:9092 && \
+                        kafka-topics --create --if-not-exists --topic platform.inventory.host-ingress-p1 --bootstrap-server kafka:9092 && \
+                        kafka-topics --create --if-not-exists --topic platform.inventory.system-profile --bootstrap-server kafka:9092 && \
+                        kafka-topics --create --if-not-exists --topic platform.platform.notifications.ingress --bootstrap-server kafka:9092 && \
+                        kafka-topics --create --if-not-exists --topic platform.inventory.events --bootstrap-server kafka:9092 && \
+                        kafka-topics --create --if-not-exists --topic platform.payload-status --bootstrap-server kafka:9092'"
+      depends_on:
+        - kafka
```
[the remark this answers is not in the document]
Comment (chambridge) [reply dev.yml:49]: I had run into kakfa issues and went digging into other items. Ultimately I think the issue was updating my systems `/etc/hosts` file with kafka being mapped to get rid of the problem. I can remove this, but thought it wasn't a bad thing to have the topics created on startup.

Flexible either way. Left it in to get the opinion from others whether there was value here.

File: tests/test_tags.py (hunk 1 above)
Comment (chambridge) [reply tests/test_tags.py:360]: I can change this, but i think this line was just copy/paste from an earlier test that was done using xjoin.

File: dev.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (kruai) [reply dev.yml:49]: I think it's fine; doesn't seem to mess up my setup. @thearifismail any objections?

**Review by kruai — DISMISSED** [review]

**Comment by thearifismail** [comment]
@kruai I don't but I can't run locally.  The unleash server can not connect to the DB container, getting blocked by ECONNREFUSED.

**Comment by chambridge** [comment]
@kruai Can you review again? I was working on the tags endpoint (next PR) and realized that the tags coming back from xjoin seem to be formatted as:
```
{"namespace": "insights-client", "key": "env", "value": "prod"} 
```
and they are stored differently in the database for example:
```
{"NS1": {"key3": ["val3"]}, "NS3": {"key3": ["val3"]}, "Sat": {"prod": []}, "SPECIAL": {"key": ["val"]}}
```
so tweaked the above to expand these into the above format: https://github.com/RedHatInsights/insights-host-inventory/pull/1596/files#diff-6e8a05b37a30d3b766362c99e6ae17fbe439a1e14e1e8f1a18695115d01c7da7R71

**Comment by kruai** [comment]
> @kruai Can you review again? I was working on the tags endpoint (next PR) and realized that the tags coming back from xjoin seem to be formatted as:
> 
> ```
> {"namespace": "insights-client", "key": "env", "value": "prod"} 
> ```
> 
> and they are stored differently in the database for example:
> 
> ```
> {"NS1": {"key3": ["val3"]}, "NS3": {"key3": ["val3"]}, "Sat": {"prod": []}, "SPECIAL": {"key": ["val"]}}
> ```
> 
> so tweaked the above to expand these into the above format: https://github.com/RedHatInsights/insights-host-inventory/pull/1596/files#diff-6e8a05b37a30d3b766362c99e6ae17fbe439a1e14e1e8f1a18695115d01c7da7R71

Actually, my normal setup doesn't work this time, leading me to think I had accidentally left a previous docker image running last time I reviewed. I'm gonna play around with it in a bit and see if I can figure out what I'm missing.

File: dev.yml [hunk 3]
```diff
@@ -6,11 +6,17 @@ services:
...
             POSTGRES_USER: insights
+            PGUSER: insights
             POSTGRES_DB: insights
         ports:
             - "5432:5432"
         volumes:
           - ./.unleash/create_unleashdb.sql:/docker-entrypoint-initdb.d/create_unleashdb.sql
+        healthcheck:
+          test: "pg_isready -h db"
+          interval: 5s
```
Comment (kruai) [inline dev.yml:17]:
```suggestion
          interval: 2s
```
Just to make startup a bit quicker

**Review by kruai — APPROVED** [review]
I just have one minor suggestion, but LGTM either way. @thearifismail please make sure this works on your setup as well before we merge

**Review by kruai — APPROVED** [review]

**Comment by thearifismail** [comment]
I need to test a bit more.  My first attempt to start the containers encountered an error stating  "unleash container unhealthy".  Stopping the containers and restarting got me past that problem.  Following the habit of running `pytest`, I executed `pytest` and failed 37 tests, mostly related to `assignment_rules`.

**Review by thearifismail — APPROVED** [review]
LGTM

File: api/host.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 from api.host_query import build_paginated_host_list_response
 from api.host_query import staleness_timestamps
 from api.host_query_db import get_all_hosts
+from api.host_query_db import get_host_tags_list_by_id_list as get_host_tags_list_by_id_list_postgres
```
[the remark this answers is not in the document]
Comment (chambridge) [reply api/host.py:16]: could certainly change this (although we are only using postgres), but if we succeed in removing xjoin then we would remove the alias here, so its temporary.

**Pull request merged**